common/flatpak-instance.c

```diff
@@ -720,6 +982,63 @@ flatpak_instance_gc_per_app_dirs (const char *instance_id,
   if (statbuf.st_mtime + 3 >= time (NULL))
     return glnx_throw (error, "lock file too recent, avoiding race condition");
 
+  /* /dev/shm is offloaded onto the host's /dev/shm to get consistent
+   * free space behaviour and make sure it's actually in RAM. It could
+   * contain relatively large files, so we clean it up.
+   *
+   * In principle this could be used for other directories such as /tmp,
+   * in a loop over an array of paths (hence this indentation), but we
+   * only do this for /dev/shm right now. */
+  do
+    {
+      g_autofree char *path = NULL;
+      glnx_autofd int lock_fd = -1;
+
+      /* /dev/shm is an attacker-controlled namespace, so we need to be
+       * careful what directories we will delete. We have to assume
+       * that attackers will create malicious symlinks in /dev/shm to
+       * try to trick us into opening or deleting the wrong files. */
+      if (flatpak_instance_claim_per_app_temp_directory (app_id,
+                                                         per_app_dir_fd,
+                                                         "dev-shm",
+                                                         "/dev/shm",
+                                                         F_WRLCK,
+                                                         &lock_fd,
+                                                         &path,
+                                                         &local_error))
+        {
+          g_assert (g_str_has_prefix (path, "/dev/shm/"));
+
+          if (!glnx_shutil_rm_rf_at (AT_FDCWD, path, NULL, &local_error))
```

Comment:
There seems to be a race condition here where someone could potentially reuse the directory in the middle of the rm, if the lock file is removed. You could make this atomic (ish) by renaming the directory before starting to remove it.

Reply:
Thanks, I'll look into this. I think it *might* actually be OK, but it's certainly subtle, and I might have missed something.

I'm really starting to think that a lock file in a directory that protects that directory is an anti-pattern, which is why I stopped using it for #4093.

Reply:
Sorry, no, I considered stopping using the lock-file-in-directory pattern in #4093 (one version had `.flatpak/.ref-com.example.App` and `.flatpak/com.example.App/`), but in the end I'm still using it there, and I address the race condition by just not deleting the directory or the lock file.

Reply:
I think this isn't actually a problem, for two orthogonal reasons (but I've fixed some different issues while I was looking at this).

The first reason it can't happen in practice is that because by the time we work with the `dev-shm` symlink, we are already holding a lock on `$XDG_RUNTIME_DIR/.flatpak/$FLATPAK_ID/.ref` (which we never delete). Because we check that the directory in `/dev/shm` has the `$FLATPAK_ID` in its name, we can know that we're holding the correct lock to exclude other processes that might be interested in it. We hold an exclusive lock (`F_WRLCK`) if we are considering deleting things, or a non-exclusive lock that conflicts with exclusive locks (`F_RDLCK`) otherwise.

However, even if that wasn't the case, I think the way we work with `.flatpak-tmpdir` would be enough to avoid the problem. The places where we take a lock on `.flatpak-tmpdir` are:

* `flatpak_instance_ensure_per_app_dev_shm()` when it creates a new directory
* `flatpak_instance_claim_per_app_temp_directory()` called by `flatpak_instance_ensure_per_app_dev_shm()` when it reuses an existing directory
* `flatpak_instance_claim_per_app_temp_directory()` called by `flatpak_instance_gc_per_app_dirs()` when it removes an existing directory

In the "newly created" code path of `flatpak_instance_ensure_per_app_dev_shm()`, the directory containing `.flatpak-tmpdir` is newly-created by `g_mkdtemp()`, so we can only be racing with another process in the very unlikely situation where `g_mkdtemp()` coincidentally gives us a directory name that matches a pre-existing symlink in `$XDG_RUNTIME_DIR`. This is also the only place where we use `O_CREAT` to create a new lock file. Ideally I would have used `O_EXCL` and `O_NOFOLLOW` here; I've added those, and also `O_NOCTTY`, for completeness.

In both the `flatpak_instance_claim_per_app_temp_directory()` code paths, we open the lock file without `O_CREAT`, so we will only open it successfully if the lock file was already created for us. I was missing `O_NOFOLLOW` here, which is potentially a problem, because a malicious app could delete the lock file and replace it with a symlink, which we would follow and (briefly) lock the target of - but this is mitigated by the fact that we are not using `O_CREAT`, so we will not create new empty files, just open existing files. Anyway, I've added that now (along with `O_NOCTTY` for completeness).

Your concern was about deletion. There are two things that could conceivably race with deleting the directory:

1. trying to reuse an existing directory
2. creating a new directory, with `g_mkdtemp()` coincidentally giving us the same directory we used before (again, unlikely but possible!)

Looking at scenario 1, where we are trying to reuse an existing directory, then both the processes that are racing will try to open `.flatpak-tmpdir` without `O_CREAT`. The process that tries to reuse the directory can either call `openat()` before `glnx_shutil_rm_rf_at()` unlinks `.flatpak-tmpdir`, or after.

If it's before, then both the processes involved in the race will have the same file open, and their locks conflict, so one of them will fail to get the lock and we're OK: if the process that wants to reuse the directory wins, then it is allowed to reuse the directory and GC does not happen. Conversely, if the process that wants to GC the directory wins, then it is allowed to delete the directory, and the other process will fail to reuse it (and recover by making a new directory instead).

If it's after, then the process that is trying to reuse the directory will have its call to `openat()` fail, and will not reuse the directory: instead it will create a new directory, the same as if it lost the race to get a lock. So we're OK either way.

Now scenario 2, where we are trying to create a new directory. `g_mkdtemp()` is basically going to end up as a retry loop, trying to `mkdir()` a new, unique directory matching `/dev/shm/flatpak-$FLATPAK_ID-XXXXXX` for some value of `XXXXXX`. Let's say it generates `/dev/shm/flatpak-$FLATPAK_ID-123456`, and let's say it's racing with a process that is calling `glnx_shutil_rm_rf_at()` on `/dev/shm/flatpak-$FLATPAK_ID-123456`, which basically means `unlink ("/dev/shm/flatpak-$FLATPAK_ID-123456/.flatpak-tmpdir")` followed by `rmdir ("/dev/shm/flatpak-$FLATPAK_ID-123456")`. Either `mkdir` happens before `rmdir`, or it happens after.

If `mkdir` happens before `rmdir`, then the `mkdir` fails with `EEXIST`, and `g_mkdtemp()` will retry with a different directory name, so we're OK.

If `rmdir` happens before `mkdir`, then GC has (basically) completely finished by the time `mkdir` gets called, so it's as though there was no race: GC finishes before reuse starts. The only thing that could still go wrong is that GC had not *100%* finished: I was using `unlinkat()` to clean up the symlink `$XDG_RUNTIME_DIR/.flatpak/$FLATPAK_ID/dev-shm` *after* deleting the directory that it points to. I've now changed the sequence so that I do the `unlinkat()` first.

Reply:
To make this more obviously correct, I'm going to remove the extra locking and just rely on the per-app-ID lock file.

Reply:
OK, new version seems to work. Now `.flatpak-tmpdir` is just a flag file to say "this is probably OK to delete", and not a lock.

Concurrent processes are guarded against by the per-app-ID lock file `$XDG_RUNTIME_DIR/.flatpak/$FLATPAK_ID/.ref`. Running apps hold a non-exclusive (read) lock, waiting for it to be available if necessary. Garbage collection takes an exclusive (write) lock, and does not wait.